src/inference/gibbs.jl

```diff
@@ -10,6 +10,11 @@ Determine whether algorithm `alg` is allowed as a Gibbs component.
 """
 isgibbscomponent(alg) = false
 
+isgibbscomponent(::ESS) = true
+isgibbscomponent(::GibbsConditional) = true
+isgibbscomponent(::Hamiltonian) = true
```

Comment:
I am wondering if we should be more restrictive here, to avoid any surprises for someone who implements another `Hamiltonian` algorithm but does not define `gibbs_state`, i.e., whose implementation is not "Gibbs-ready".

Reply:
I agree.  And beyond that: since the "Gibbs-readyness" is a trait implementation, why move most of the code away from the individual samplers, and into this file?  Especially since we are moving away from priviledged implementations here.

IMHO things are clearer to look up when the Gibbs-parts are also near them.  E.g., from the perspective of someone looking up how to implement a Gibbs component.

Reply:
> why move most of the code away from the individual samplers, and into this file?

I agree that this would be better, and in fact it was the case initially in this PR. However, I noticed that it does not work for `gibbs_rerun` - Turing can't be compiled anymore if `gibbs_rerun` is defined next to the corresponding sampler due to the dependency on two samplers, e.g., `gibbs_rerun(::MH, ::Hamiltonian)` requires `MH` to be defined before Hamiltonian and similarly `gibbs_rerun(::Hamiltonian, ::MH)` requires `Hamiltonian` to be defined before `MH`. Therefore I decided it is more consistent to move all Gibbs parts for the Turing-internal samplers to the `Gibbs` implementation - otherwise some traits would be implemented in `gibbs.jl` and some in the files of the corresponding samplers. Only DynamicNUTS is an exception here and treated like a user-defined sampler that has to bundle also the Gibbs parts since it is loaded optionally if DynamicHMC is available.

Reply:
Uh, I see.  An unfortunate situation.  But how does the two-argument `gibbs_rerun` generalize between user-defined samplers?  

In a way this makes me more sympathetic to the old way.  Was the meaning of `selector.rerun` exactly "the logp has not been calculated", or was there more logic to it?

Reply:
There's much more to it: `vi.logp` means different things for different samplers (sometimes it is the joint log probability, sometimes it is the likelihood, sometimes it is something in between, and sometimes it is something completely different), some samplers just don't care about the value of `vi.logp` in their sampling step or reset everything to zero, etc.

Therefore I think the only reasonable way to define `rerun` (regardless how it is saved or implemented) is to compare two concrete samplers. The default setting of `gibbs_rerun(prev_sampler, sampler) = true` makes sure that every user implementation is always mathematically correct, and then users have to explicitly define if one can save one computation for a specific previous sampler by implementing `gibbs_rerun` for this specific sampler. The current implementation is just wrong, to be honest, since it ignores all these facts - however, I assume it does not lead to any problems since `vi.logp` is basically always recomputed, apart from `MH` with a previous sampler that is neither `ESS`, `PG`, nor `GibbsConditional`. :shrug: